90% PLAT-1871 Check HTTP Status Codes for non 200 responses.

malcyL opened the pull request:
#### Problem

babel-server is returning a 400 response for requests, but this is not being translated into an error in the client. 

This client is checking the body of the response for the json property errors. The response reported in PLAT-1871 contained the error in an error property rather than an errors property.

#### Fix

We could change the node client to check for error. Would we need it to check for errors too?

The correct solution here is to listen to HTTP response codes rather than looking for error or errors in the response body.

This PR makes the node client error if the response code is not the 2xx range.

I did consider completely removing the existing code checking for the existence of errors in the body. However, without checking all responses are correctly using error codes, I'm reluctant to do that. If we have been checking for errors using this mechanism, it's possible we are returning errors without using the correct status code.

For now, I'm adding in a new check for the status code, but leaving in the existing code to ensure there is no unexpected behavior from this change. At some point, we should audit all responses from babel-server and if it's ok, remove the checking for the existence of errors in the body completely. But not now.

malcyL commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+*.swp
```
Vim for the win! ;-)

russellhill commented on index.js:
```diff
@@ -607,6 +607,15 @@ BabelClient.prototype.error = function error(message) {
     this._log(ERROR, message);
 };
 
+BabelClient.prototype.responseSuccessful = function responseSuccessful(response) {
```
should this not be `_responseSuccessful` to try and indicate it's an internal "private" method?

markwallsgrove replied:
I would suggest changing the names would be out of scope, but we should refactor it in the future.

russellhill commented on index.js:
```diff
@@ -364,15 +364,15 @@ BabelClient.prototype.createAnnotation = function createAnnotation(token, data,
             callback(err);
         } else{
-            if(body.message && body.errors){
+            if(!this.responseSuccessful(response) || (body.message && body.errors)){
```
I thought we determined it was returning `.error` as part of the `body` and not `.errors`?

malcyL replied:
I've not changed what the existing code is doing here re error/errors - I've just added a check for the status code.

The client is checking for 'errors' - but at least some responses were returning 'error' as you say. This must always have been broken. But that approach is flawed - as evidenced by this bug. I don't think we should change the error/errors checking, because I'm not sure of all the possible responses from babel. But I am sure that if we return a status code of 400, it should not be a successful request.

markwallsgrove commented:
just seen another bit of funky code in the babel node client

`} else if(response.statusCode && response.statusCode !== 204){`

the `response.statusCode` might be zero, which occurs on network failures. This would result in a success being sent back

there are more of these too

```js
  request(requestOptions, function requestResponse(err, response, body){
    if(err){
      callback(err);
    } else{
      self._parseJSON(response, body, callback);
    }
  });
```

malcyL commented:
@markwallsgrove I don't believe that's right...

If there is a network error etc, the request function will return an error and we will not even reach the code I've changed:

```
    request.post(requestOptions, function requestResponse(err, response, body){
        if(err){
            callback(err);  <<<<<========== Network error dealt with as before here
        } else{
            if(!this._responseSuccessful(response) || (body.message && body.errors)){
                var babelError = new Error(body.message);
                babelError.http_code = response.statusCode || 404;
                callback(babelError);
            } else{
                callback(null, body);
            }
        }
    }.bind(this));
}
```

We'll only get to the code that I've changed if we make a HTTP request and get a response, i.e. there was no error making the http request itself. In that scenario we will have a staus code. I'm not convinced we would get a status code of zero at that point - but even if we did, the success code is checking for a code in the 2xx range, so we would not treat a zero response code as success:

```
BabelClient.prototype._responseSuccessful = function responseSuccessful(response) {
    if (response && response.statusCode) {
        if (response.statusCode >= 200 && response.statusCode <= 299) {
            return true;
        }
    }
    return false;
};
```

markwallsgrove commented:
@malcyL I was referring to this code
```js
} else if(response.statusCode && response.statusCode !== 204){
```
which isn't a change you have made, but it is within that file.

The `else if` statement above is checking if a error should be returned. The first clause is `response.statusCode`, which will result in a boolean false if the value is zero. The code was expecting to  check if the attribute existed, not the value.

The code you have above would correctly assert the state as it is checking for a success state, not a failure state.

markwallsgrove commented:
@malcyL the latter

```js
  request(requestOptions, function requestResponse(err, response, body){
    if(err){
      callback(err);
    } else{
      self._parseJSON(response, body, callback);
    }
  });
```
I was suggesting checking the response status also, like what you did above.

malcyL commented:
@markwallsgrove I want to keep this PR just on the error that @russellhill raised.

We are checking for the existence of "error/errors" in the response body in two places - and my code addresses both of those. I don't really want to expand this fix beyond that specific bug. We can raise other issues and PR's for further fixes.

malcyL commented:
@russellhill I'd like to get this merged for our current sprint. Re my comment:

> However, would it be an idea for you to build a version of TN using this branch of the client so we know it is fixing your problem before I merge and release?

This is passing our tests so I would be happt to merge and release if you don't have time to do this?
